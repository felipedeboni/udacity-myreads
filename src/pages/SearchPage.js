import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import { debounce } from 'throttle-debounce';
import * as BooksAPI from '../utils/BooksAPI';
import Book from '../components/Book';

class SearchPage extends Component {
  constructor() {
    super();

    this.search = debounce(400, this.search);

    this.state = {
      query: '',
      results: []
    };
  }

  updateQuery = query => {
    this.setState({ query: query });

    if (!query.trim()) {
      this.clear();
    } else {
      this.search(query);
    }
  };

  search = query => {
    // we need to check if the query is equals to the state query bc the debounce
    if (this.state.query !== query) return;

    BooksAPI.search(this.state.query, 10).then(results => {
      if (results.error) {
        return this.clear();
      }
      this.setState({ results });
    });
  };

  clear = () => {
    this.setState({
      results: []
    });
  };

  render() {
    const { myBooks, onMoveToShelf } = this.props;

    const results = this.state.results.map(book => {
      const mybook = myBooks.find(_ => _.id === book.id) || {};
      book.shelf = !!mybook.shelf ? mybook.shelf : 'none';

      return book;
    });

    return (
      <div className="search-books">
        <div className="search-books-bar">
          <Link to="/" className="close-search">
            Close
          </Link>
          <div className="search-books-input-wrapper">
            {/*
              NOTES: The search from BooksAPI is limited to a particular set of search terms.
              You can find these search terms here:
              https://github.com/udacity/reactnd-project-myreads-starter/blob/master/SEARCH_TERMS.md

              However, remember that the BooksAPI.search method DOES search by title or author. So, don't worry if
              you don't find a specific author or title. Every search is limited by search terms.
            */}
            <input
              type="text"
              placeholder="Search by title or author"
              value={this.state.query}
              onChange={event => this.updateQuery(event.target.value)}
            />
          </div>
        </div>
        <div className="search-books-results">
          <ol className="books-grid">
            {results.map(book => (
              <li key={book.id}>
                <Book
                  book={book}
                  hideNone={true}
                  onMoveToShelf={onMoveToShelf}
                />
              </li>
            ))}
          </ol>
        </div>
      </div>
    );
  }
}

SearchPage.propTypes = {
  myBooks: PropTypes.array.isRequired,
  onMoveToShelf: PropTypes.func.isRequired
};

export default SearchPage;
