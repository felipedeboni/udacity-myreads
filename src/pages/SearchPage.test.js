import React from 'react';
import SearchPage from './SearchPage';
import { shallow } from 'enzyme';

const results = [
  {
    title: 'React',
    subtitle: 'Die praktische Einführung in React, React Router und Redux',
    authors: ['Nils Hartmann', 'Oliver Zeigermann'],
    publisher: 'dpunkt.verlag',
    publishedDate: '2016-07-07',
    description:
      'React ist ein JavaScript-Framework zur Entwicklung von Benutzeroberflächen sowohl im Browser als auch auf Mobilgeräten. Entwickelt und eingesetzt von Facebook ist es mittlerweile als Open-Source-Projekt verfügbar und hat sich bereits im Einsatz bei diversen namhaften Websites, wie z. B. Airbnb und Netflix, bewährt. Dieses Buch stellt Ihnen die Konzepte von React, React Router und Redux anhand eines durchgehenden Beispiels vor. Sie lernen, wie Sie mit React wiederverwendbare UI-Komponenten entwickeln und wie Sie auf Basis der einzelnen Komponenten ganze Anwendungen zusammenbauen. Unter anderem werden folgende Themen behandelt: - Entwickeln und Testen eigener React-Komponenten auf Basis des JavaScript-Standards ECMAScript 2015 (ES6) - Routing mit dem React Router - Das Architektur-Modell Flux und wie damit komplette Anwendungen umgesetzt werden (am Beispiel des Redux-Frameworks) - Serverseitiges Rendern von React-Komponenten und -Anwendungen - Anbindung eines REST-Backends Die im Buch eingesetzten Sprachfeatures aus ES6 werden in einem eigenen Kapitel vorgestellt, sodass zum Verständnis des Buches Kenntnisse von ES5 ausreichen. Nach der Lektüre des Buches werden Sie in der Lage sein, eigene Projekte mit React umzusetzen.',
    industryIdentifiers: [
      {
        type: 'ISBN_13',
        identifier: '9783864919640'
      },
      {
        type: 'ISBN_10',
        identifier: '3864919649'
      }
    ],
    readingModes: {
      text: true,
      image: false
    },
    pageCount: 342,
    printType: 'BOOK',
    categories: ['Computers'],
    maturityRating: 'NOT_MATURE',
    allowAnonLogging: true,
    contentVersion: '1.4.3.0.preview.2',
    panelizationSummary: {
      containsEpubBubbles: false,
      containsImageBubbles: false
    },
    imageLinks: {
      smallThumbnail:
        'http://books.google.com/books/content?id=IOejDAAAQBAJ&printsec=frontcover&img=1&zoom=5&source=gbs_api',
      thumbnail:
        'http://books.google.com/books/content?id=IOejDAAAQBAJ&printsec=frontcover&img=1&zoom=1&source=gbs_api'
    },
    language: 'de',
    previewLink:
      'http://books.google.com/books?id=IOejDAAAQBAJ&dq=redux+react&hl=&cd=15&source=gbs_api',
    infoLink:
      'http://books.google.com/books?id=IOejDAAAQBAJ&dq=redux+react&hl=&source=gbs_api',
    canonicalVolumeLink:
      'https://books.google.com/books/about/React.html?hl=&id=IOejDAAAQBAJ',
    id: 'IOejDAAAQBAJ'
  }
];

it('renders without crashing', () => {
  const wrapper = shallow(
    <SearchPage myBooks={[]} onMoveToShelf={jest.fn()} />
  );

  expect(wrapper).toMatchSnapshot();
});

it('should include a link to the homepage', () => {
  const wrapper = shallow(
    <SearchPage myBooks={[]} onMoveToShelf={jest.fn()} />
  );

  expect(wrapper.find('Link[to="/"]').exists()).toBe(true);
});

it('should not be possible to remove book from shelf', () => {
  const wrapper = shallow(
    <SearchPage myBooks={[]} onMoveToShelf={jest.fn()} />
  );

  wrapper.setState({ results });
  expect(wrapper.find('Book[hideNone=true]').exists()).toBe(true);
});

it('clear should clear all results', () => {
  const wrapper = shallow(
    <SearchPage myBooks={[]} onMoveToShelf={jest.fn()} />
  );

  wrapper.setState({ results });
  wrapper.instance().clear();
  wrapper.update();
  expect(wrapper.find('Book').length).toBe(0);
});

it('update query should call search when query is valid', () => {
  const wrapper = shallow(
    <SearchPage myBooks={[]} onMoveToShelf={jest.fn()} />
  );

  const inst = wrapper.instance();
  const searchFn = jest.fn();

  inst.search = searchFn;
  inst.updateQuery('android');
  expect(searchFn).toHaveBeenCalled();
});

it('update query should call clear when query is invalid', () => {
  const wrapper = shallow(
    <SearchPage myBooks={[]} onMoveToShelf={jest.fn()} />
  );

  const inst = wrapper.instance();
  const clearFn = jest.fn();

  inst.clear = clearFn;
  inst.updateQuery(' ');
  expect(clearFn).toHaveBeenCalled();
});

it('search input should call update query when changed', () => {
  const wrapper = shallow(
    <SearchPage myBooks={[]} onMoveToShelf={jest.fn()} />
  );

  const inst = wrapper.instance();
  const updateQueryFn = jest.fn();

  inst.updateQuery = updateQueryFn;
  wrapper.find('.search-books-input-wrapper input').simulate('change', {
    target: {
      value: 'android'
    }
  });

  expect(updateQueryFn).toHaveBeenCalled();
});
