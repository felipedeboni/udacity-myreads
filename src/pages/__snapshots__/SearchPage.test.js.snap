// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without crashing 1`] = `
<div
  className="search-books"
>
  <div
    className="search-books-bar"
  >
    <Link
      className="close-search"
      replace={false}
      to="/"
    >
      Close
    </Link>
    <div
      className="search-books-input-wrapper"
    >
      <input
        onChange={[Function]}
        placeholder="Search by title or author"
        type="text"
        value=""
      />
    </div>
  </div>
  <div
    className="search-books-results"
  >
    <ol
      className="books-grid"
    />
  </div>
</div>
`;
